conformance/utils/kubernetes/apply.go

```diff
@@ -41,45 +42,95 @@ import (
 // Applier prepares manifests depending on the available options and applies
 // them to the Kubernetes cluster.
 type Applier struct {
-	NamespaceLabels map[string]string
-	// ValidUniqueListenerPorts maps each listener port of each Gateway in the
-	// manifests to a valid, unique port. There must be as many
-	// ValidUniqueListenerPorts as there are listeners in the set of manifests.
-	// For example, given two Gateways, each with 2 listeners, there should be
-	// four ValidUniqueListenerPorts.
-	// If empty or nil, ports are not modified.
-	ValidUniqueListenerPorts []v1alpha2.PortNumber
+	namespaceLabels map[string]string
+	// availablePorts holds ports that can still be assigned
+	availablePorts PortSet
+	// assignedPorts tracks ports used for Gateway objects
+	assignedPorts map[types.NamespacedName]PortSet
+	portsLock     sync.Mutex
+}
+
+// NewApplier creates a new Applier object.
+// validPorts should contain a list of valid ports for Gateway listeners.
+// There must be as many validPorts as the maximum number of ports
+// used simultaneously.
+// For example, given one Gateway with 2 listeners on the same port and one
+// Gateway with 2 listeners on different ports, there should be at least three
+// validPorts.
+// If empty or nil, Gateway listener ports are not modified.
+func NewApplier(namespaceLabels map[string]string, validPorts []v1alpha2.PortNumber) *Applier {
+	return &Applier{
+		namespaceLabels: namespaceLabels,
+		availablePorts:  validPorts,
+		assignedPorts:   map[types.NamespacedName]PortSet{},
+	}
+}
+
+// freePortsFor is used to mark ports used by a Gateway as available. It does
+// not lock the availablePorts map.
+func (a *Applier) freePortsFor(name types.NamespacedName) {
+	ports, ok := a.assignedPorts[name]
+	if !ok {
+		return
+	}
+	for _, port := range ports {
+		a.availablePorts = append(a.availablePorts, port)
+	}
+	delete(a.assignedPorts, name)
 }
 
 // prepareGateway adjusts both listener ports and the gatewayClassName. It
-// returns an index pointing to the next valid listener port.
-func prepareGateway(t *testing.T, uObj *unstructured.Unstructured, gatewayClassName string, validListenerPorts []v1alpha2.PortNumber, portIndex int) int {
+// returns the ports used by the listeners if they came from validPorts.
+func (a *Applier) prepareGateway(t *testing.T, uObj *unstructured.Unstructured, gatewayClassName string) {
```

Comment:
Is the idea that this function can only be called if portsLock has already been locked?

Reply:
Yes.